### PR TITLE
Use grayscale AA always on non-opaque backgrounds

### LINKED ISSUE #5098: Using cleartype anti aliasing makes dark foreground invisible on light background.
# Environment

```none
Windows build number: 10.0.18363.0
Windows Terminal version: 0.10.781.0
Powershell version: 7.0.0
```

# Steps to reproduce
Setting profile settings to following:
```json
{
    "useAcrylic": false,
    "closeOnExit" : true,
    "colorScheme" : "Campbell",
    "guid": "{a3c89eef-1dbe-4ac1-acc5-2412499f8619}",
    "hidden": false,
    "name": "PowerShell 7",
    "commandline" : "C:\\Program Files\\PowerShell\\7\\pwsh.exe",
    "icon" : "C:\\Program Files\\PowerShell\\7\\assets\\Powershell_av_colors.ico",
    "antialiasingMode": "grayscale",
    "cursorShape" : "bar",
    "cursorColor": "#121005",
    "fontFace" : "Consolas",
    "fontSize" : 10,
    "snapOnInput" : true,
    "historySize" : 9001
}
```
and theme settings to:
```json
{
    "name" : "Campbell",

    "foreground" : "#020200",
    "background" : "#ffffff",
    "selectionBackground": "#121005",

    "black" : "#0C0C0C",
    "blue" : "#0037DA",
    "brightBlack" : "#767676",
    "brightBlue" : "#3B78FF",
    "brightCyan" : "#61D6D6",
    "brightGreen" : "#16C60C",
    "brightPurple" : "#B4009E",
    "brightRed" : "#E74856",
    "brightWhite" : "#121212",
    "brightYellow" : "#3B78FF",
    "cyan" : "#3A96DD",
    "green" : "#13A10E",
    "purple" : "#881798",
    "red" : "#C50F1F",
    "white" : "#CCCCCC",
    "yellow" : "#C19C00"
}
```

# Expected behavior

Black text `F:\>` is visible. When antialiasingMode is set to cleartype.

# Actual behavior

Black text `F:\>` is invisible after setting antialiasingMode to cleartype (on the right):
image
Same issue with cmd ("acrylicOpacity" : 0.75):
image

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

When we're on acrylic, we can't have cleartype text unfortunately. This PR changes the DX renderer to force cleartype runs of text that are on a non-opaque background to use grayscale AA instead.

## References

Here are some of the URLS I was referencing as writing this:

* https://stackoverflow.com/q/23587787
* https://docs.microsoft.com/en-us/windows/win32/direct2d/supported-pixel-formats-and-alpha-modes#cleartype-and-alpha-modes
* https://devblogs.microsoft.com/oldnewthing/20150129-00/?p=44803
* https://docs.microsoft.com/en-us/windows/win32/api/d2d1/ne-d2d1-d2d1_layer_options
* https://docs.microsoft.com/en-us/windows/win32/direct2d/direct2d-layers-overview#d2d1_layer_parameters1-and-d2d1_layer_options1
* https://docs.microsoft.com/en-us/windows/win32/api/dcommon/ne-dcommon-d2d1_alpha_mode?redirectedfrom=MSDN#cleartype-and-alpha-modes
* https://stackoverflow.com/a/26523006

Additionally:
* This was introduced in #4711 

## PR Checklist
* [x] Closes #5098
* [x] I work here
* [ ] Tests added/passed
* [n/a] Requires documentation to be updated

## Detailed Description of the Pull Request / Additional comments

Basically, if you use cleartype on a light background, what you'll get today is the text foreground color _added_ to the background. This will make the text look basically invisible. 

So, what I did was use some trickery with `PushLayer` to basically create a layer where the text would be forced to render in grayscale AA. I only enable this layer pushing business when both:
  * The user has enabled cleartype text
  * The background opacity < 1.0

This plumbs some information through from the TermControl to the DX Renderer to make this smooth.

## Validation Steps Performed

Opened both cleartype and grayscale panes SxS, and messed with the opacity liberally.

### CODE REVIEW

File: src/renderer/dx/CustomTextRenderer.cpp [hunk 1]
```diff
@@ -286,6 +286,39 @@ using namespace Microsoft::Console::Render;
...
+                                                    D2D1::IdentityMatrix(),
+                                                    1.0,
+                                                    nullptr,
+                                                    D2D1_LAYER_OPTIONS_INITIALIZE_FOR_CLEARTYPE),
+                              nullptr);
+    }
+    else
+    {
+        popLayer.release();
+    }
```
Comment (miniksa) [inline src/renderer/dx/CustomTextRenderer.cpp:321]: We have to ensure that we're only doing this for the specific color-AA (ClearType) on the acrylic background as adding a layer is moderate performance detriment.

We should also leave behind the comment that says DANGER: Layers slow us down. Only do this in the specific case where someone has chosen the slower ClearType antialiasing (versus the faster grayscale antialiasing).

**Comment by zadjii-msft** [comment]
**BLOCK THIS** 

It doesn't work on first launch when you've got acrylic + cleartype. In that scenario it starts by using the solid BG by default. Need to investigate...

**Comment by zadjii-msft** [comment]
> **BLOCK THIS**
> 
> It doesn't work on first launch when you've got acrylic + cleartype. In that scenario it starts by using the solid BG by default. Need to investigate...

Fixed in latest commit

File: src/renderer/dx/DxRenderer.cpp [hunk 2]
```diff
@@ -2130,3 +2168,23 @@ void DxEngine::SetAntialiasingMode(const D2D1_TEXT_ANTIALIAS_MODE antialiasingMo
...
+// - opacity: the new opacity of our background, on [0.0f, 1.0f]
+// Return Value:
+// - <none>
+void DxEngine::SetDefaultTextBackgroundOpacity(const float opacity) noexcept
+{
+    _defaultTextBackgroundOpacity = opacity;
+
+    // Make sure we redraw all the cells, to update whether they're actually
+    // drawn with cleartype or not.
+    // We don't terribly care if this fails.
```
Comment (carlos-zamora) [inline src/renderer/dx/DxRenderer.cpp:2188]: Just curious, why don't we care?

Comment (miniksa) [reply src/renderer/dx/DxRenderer.cpp:2188]: It's not worth crashing over if it does fail. It'll just make a graphical glitch... is my most likely candidate for a reason.

Comment (zadjii-msft) [reply src/renderer/dx/DxRenderer.cpp:2188]: that's definitely the reason

**Comment by zadjii-msft** [comment]
@DHowett-MSFT you blocked this for me, unblock yo self plz

**Comment by DHowett-MSFT** [comment]
Oop
